task_manager/task_manager.hh

```diff
@@ -0,0 +1,229 @@
+        foreign_task_vector _children;
+        seastar::shared_promise<> _done;
+    public:
+        task(task_id id, foreign_task_ptr&& parent = nullptr) {
+            _status.id = id;
+            _status.state = task_state::created;
+            _status.shard = this_shard_id();
+            if (parent.get() != nullptr) {
+                _parent = std::make_unique<foreign_task_ptr>(std::move(parent));
+            }
+        }
+
+        status& get_status() {
+            return _status;
+        }
+
+        bool has_parent() const {
+            return _parent != nullptr;
+        }
+
+        // Can be called only if the task has a parent task.
+        task_id get_parent_id() const {
+            return _parent->get()->_status.id;
+        }
+
+        // Can be called only if the task has a parent task.
+        foreign_task_ptr& get_parent() {
+            return *_parent;
+        }
+
+        future<progress> get_progress() {
+            if (_children.empty()) {
+                return make_ready_future<progress>(_status.progress);
+            }
+
+            return map_reduce(_children, [] (auto& child) {
+                return smp::submit_to(child.get_owner_shard(), [&child] {
+                    return child->get_progress();
+                });
+            }, progress{}, [] (progress l, progress r) {
+                progress res{};
+                res.total = l.total + r.total;
+                res.completed = l.completed + r.completed;
+                return res;
+            });
+
+        }
+
+        virtual std::string get_module() const = 0;
+
+        virtual is_abortable is_abortable() const {
+            return is_abortable::no;
+        }
+
+        void set_number(int number) {
+            _status.number = number;
+        }
+
+        void set_type(std::string type) {
+            _status.type = type;
+        }
+
+        void change_state(task_state state) {
+            _status.state = state;
+        }
+
+        void set_keyspace(std::string keyspace) {
+            _status.keyspace = keyspace;
+        }
+
+        void set_table(std::string table) {
+            _status.table = table;
+        }
+
+        void set_entity(std::string entity) {
+            _status.entity = entity;
+        }
+
+        void set_total_progress(double progress) {
+            _status.progress.total = progress;
+        }
+
+        void report_completion_progress(double progress) {
+            _status.progress.completed += progress;
+        }
+
+        void start(task_manager& tm) {
+            _status.start_time = db_clock::now();
+            _status.state = task_manager::task_state::running;
+
+            (void)with_gate(tm.async_gate(), [f = done(), &tm, id = _status.id] () mutable {
+                return std::move(f).finally([&tm, id] {
+                    return sleep_abortable(tm.task_ttl(), tm.abort_source());
+                }).finally([&tm, id] {
```

Comment:
sleep_abortable throws `sleep_aborted` exception when aborted.
To ignore it we better use `.then_wrapped` here and call `f.ignore_ready_future()` to ignore the exception.
Otherwise, seastar will complain about ignored exception when this chain will be destroyed.